### PR TITLE
Check SerializableLambda parameters for prohibited types in CC storing

### PR DESCRIPTION
**Pull request by h0tk3y**
Fixes https://github.com/gradle/gradle/issues/17821

### CODE REVIEW

**Comment by h0tk3y** [comment]
@bot-gradle test ReadyForNightly

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/SerializedLambdaParametersCheckingCodec.kt [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ * If some of them are not, it logs a problem.
+ *
+ * The lambda is encoded straightforwardly as a [BeanSpec], and, upon decoding, a [BeanSpec] is expected to contain the [SerializedLambda].
+ * Beside the compliance checks, the values are encoded or decoded as beans without any special handling.
+ *
+ * @see [org.gradle.configurationcache.serialization.beans.unsupportedFieldDeclaredTypes]
+ */
+object SerializedLambdaParametersCheckingCodec : Codec<SerializedLambda> {
+    override suspend fun ReadContext.decode(): SerializedLambda {
+        return readNonNull<BeanSpec>().bean as SerializedLambda
```
Comment (bamboo) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/SerializedLambdaParametersCheckingCodec.kt:44]: 🚀 Using `decodeBean()` here would be slightly faster (avoid a recursion roundtrip) and simpler.

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/SerializedLambdaParametersCheckingCodec.kt [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    override suspend fun WriteContext.encode(value: SerializedLambda) {
+        checkLambdaCapturedArgTypesAreSupported(value)
+        write(BeanSpec(value))
```
Comment (bamboo) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/SerializedLambdaParametersCheckingCodec.kt:49]: 🚀 Using `encodeBean(value)` here would be slightly faster an simpler.

**Review by bamboo — CHANGES_REQUESTED** [review]

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/SerializedLambdaParametersCheckingCodec.kt (hunk 1 above)
Comment (h0tk3y) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/SerializedLambdaParametersCheckingCodec.kt]: Done in https://github.com/gradle/gradle/pull/23156/commits/aedb6804f13ee945db7ec1018eac87dbf7b963f9

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/SerializedLambdaParametersCheckingCodec.kt (hunk 2 above)
Comment (h0tk3y) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/SerializedLambdaParametersCheckingCodec.kt]: Done in https://github.com/gradle/gradle/pull/23156/commits/aedb6804f13ee945db7ec1018eac87dbf7b963f9

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/SerializedLambdaParametersCheckingCodec.kt [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        checkLambdaCapturedArgTypesAreSupported(value)
+        encodeBean(value)
+    }
+
+    private
+    fun IsolateContext.checkLambdaCapturedArgTypesAreSupported(value: SerializedLambda) {
+        val signature = value.implMethodSignature
+        val paramTypes: Array<Type> = getArgumentTypes(signature)
+
+        // Treat all parameters equally, regardless of whether they are implicit captured parameters or the lambda signature ones.
```
Comment (bamboo) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/SerializedLambdaParametersCheckingCodec.kt:58]: 👍

**Review by bamboo — APPROVED** [review]
:tada:

**Comment by h0tk3y** [comment]
@bot-gradle test and merge

**Pull request merged**